### PR TITLE
feat: updated I2C API and fixed BMP180

### PR DESCRIPTION
**Pull request by AsCress**
**Updates** the I2C API, making it akin to the one present in _pslab-python_.
**Fixes** the library for the `BMP180` Altimeter.

## Changes 
- The I2C API is updated to add `I2CSlave` functions such as `read`, `write`, `readInt`, etc., similar to the ones present in _pslab-python_.
- The library implementation of the _BMP180_ Altimeter is now fixed in Android as well (refer https://github.com/fossasia/pslab-python/pull/231).

## Screenshots / Recordings  
Here is a screen recording showing the BMP180 interface. The readings are consistent with the results provided by other _python_ and _Arduino_ drivers.

https://github.com/user-attachments/assets/ee3ce3d0-7636-43e0-ba7c-b3d016ff375d

**Checklist**: 
- [x] **No hard coding**: I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard coding any value.
- [x] **No end of file edits**: No modifications done at end of resource files `strings.xml`, `dimens.xml` or `colors.xml`.
- [x] **Code reformatting**: I have reformatted code and fixed indentation in every file included in this pull request.
- [x] **No extra space**: My code does not contain any extra lines or extra spaces than the ones that are necessary.

@marcnause @CloudyPadmal @bessman This is my first work on the _I2C_ implementation in Android, which sets up everything for the upcoming big additions. All aspects related to the communication are now sorted in Android. The `BMP180`  is used as a test for the same.

## Summary by Sourcery

Update the I2C API to include new functions similar to pslab-python and fix the BMP180 Altimeter library for Android. Refactor sensor code to improve readability and switch to integer-based data handling.

New Features:
- Updated I2C API to include new functions such as read, write, readInt, and others, aligning it with the pslab-python implementation.

Bug Fixes:
- Fixed the BMP180 Altimeter library implementation for Android, ensuring consistent readings.

Enhancements:
- Refactored BMP180 sensor code to improve readability and maintainability, including changes to variable names and data handling.
- Replaced character-based data handling with integer-based data handling across multiple sensor classes for better consistency and performance.

### CODE REVIEW

File: app/src/main/java/io/pslab/communication/sensors/BMP180.java [hunk 1]
```diff
@@ -19,108 +19,146 @@
...
     public String[] PLOTNAMES = {"Temperature", "Pressure", "Altitude"};
     public String name = "Altimeter BMP180";
 
     private I2C i2c;
-    private int MB;
-    private double c3, c4, b1, c5, c6, mc, md, x0, x1, x2, y0, y1, y2, p0, p1, p2, temperature, pressure, baseline;
-    private ArrayList<Integer> setOverSampling = new ArrayList<>(Arrays.asList(0, 1, 2, 3));
+    private int ac1, ac2, ac3, ac4, ac5, ac6, b1, b2, mb, mc, md;
+    private double temperature, pressure;
+    private final double seaLevelPressure = 101325.0;
```
[the remark this answers is not in the document]
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/sensors/BMP180.java]: At the sea level, the standard air pressure doesn't change at different locations. It remains the same everywhere.

File: app/src/main/java/io/pslab/communication/peripherals/I2C.java [hunk 2]
```diff
@@ -140,19 +140,37 @@ public int readStatus() throws IOException {
...
         }
-        return charData;
+        return intData;
+    }
+
+    public ArrayList<Integer> read(int deviceAddress, int bytesToRead, int registerAddress) throws IOException {
+        return readBulk(deviceAddress, registerAddress, bytesToRead);
+    }
+
+    public int readByte(int deviceAddress, int registerAddress) throws IOException {
```
[the remark this answers is not in the document]
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/peripherals/I2C.java:162]: Null checks are already present in the fundamental reading methods.

File: app/src/main/java/io/pslab/communication/sensors/BMP180.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/sensors/BMP180.java]: Not needed

File: app/src/main/java/io/pslab/communication/peripherals/I2C.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/peripherals/I2C.java:162]: Current implementation is clear enough.

**Comment by marcnause** [comment]
I get values from my BMP180 that look pretty reasonable.

The temperature is pretty accurate. The pressure is off by approximately 500 Pa from what the internet says the current pressure is where I live. The altitude is off my about 40m, but since it is calculated from the pressure, I would not expect it to be very accurate.

File: app/src/main/java/io/pslab/communication/sensors/BMP180.java
Comment (marcnause) [inline app/src/main/java/io/pslab/communication/sensors/BMP180.java:1]: I propose to make all constants in this class `static final` to fulfill Java conventions.

File: app/src/main/java/io/pslab/communication/peripherals/I2C.java [hunk 3]
```diff
@@ -140,19 +140,37 @@ public int readStatus() throws IOException {
...
         packetHandler.sendByte(registerAddress);
         packetHandler.sendByte(bytesToRead);
         byte[] data = new byte[bytesToRead + 1];
         packetHandler.read(data, bytesToRead + 1);
-        ArrayList<Character> charData = new ArrayList<>();
+        ArrayList<Integer> intData = new ArrayList<>();
         for (int i = 0; i < bytesToRead; i++) {
-            charData.add((char) data[i]);
+            intData.add((int) data[i]);
         }
```
Comment (marcnause) [inline app/src/main/java/io/pslab/communication/peripherals/I2C.java:154]: It should be possible to simplify this loop to:

`for (byte b : data) {`
`   intData.add((int) b);`
`}`

File: app/src/main/java/io/pslab/communication/sensors/BMP180.java
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/sensors/BMP180.java:1]: I've made all the constants `static final`.

File: app/src/main/java/io/pslab/communication/peripherals/I2C.java (hunk 3 above)
Comment (AsCress) [reply app/src/main/java/io/pslab/communication/peripherals/I2C.java:154]: Thanks for the suggestion ! I've simplified the loop.

**Comment by AsCress** [comment]
@marcnause Thank you very much for testing and confirming that the sensor and the _I2C_ communication works well ! The errors you mentioned are normal when working with a sensor such as the `BMP180`. I've revised the code to implement all the suggested changes.

**Review by marcnause — APPROVED** [review]

**Pull request merged**